ProcessLib/HydroMechanics/HydroMechanicsProcess-impl.h

```diff
@@ -84,146 +158,231 @@ void HydroMechanicsProcess<DisplacementDim>::initializeConcreteProcess(
+        dof_tables;
+    // For the monolithic scheme
+    if (_use_monolithic_scheme)
+    {
+        DBUG(
+            "Assemble the Jacobian of HydroMechanics for the monolithic"
+            " scheme.");
+        dof_tables.push_back(std::ref(*_local_to_global_index_map));
+    }
+    else
+    {
+        // For the staggered scheme
+        if (_coupled_solutions->process_id == 0)
+        {
+            DBUG(
+                "Assemble the Jacobian equations of liquid fluid process in "
+                "HydroMechanics for the staggered scheme.");
+        }
+        else
+        {
+            DBUG(
+                "Assemble the Jacobian equations of mechanical process in "
+                "HydroMechanics for the staggered scheme.");
+        }
+        std::vector<std::reference_wrapper<NumLib::LocalToGlobalIndexMap>>
+            dof_tables = {std::ref(*_local_to_global_index_map_with_base_nodes),
+                          std::ref(*_local_to_global_index_map)};
+    }
 
-    // Call global assembler for each local assembly item.
     GlobalExecutor::executeMemberDereferenced(
         _global_assembler, &VectorMatrixAssembler::assembleWithJacobian,
-        _local_assemblers, *_local_to_global_index_map, t, x, xdot, dxdot_dx,
-        dx_dx, M, K, b, Jac, _coupled_solutions);
+        _local_assemblers, dof_tables, t, x, xdot, dxdot_dx, dx_dx, M, K, b,
+        Jac, _coupled_solutions);
 }
+
 template <int DisplacementDim>
 void HydroMechanicsProcess<DisplacementDim>::preTimestepConcreteProcess(
     GlobalVector const& x, double const t, double const dt,
-    const int /*process_id*/)
+    const int process_id)
 {
     DBUG("PreTimestep HydroMechanicsProcess.");
 
     _process_data.dt = dt;
     _process_data.t = t;
 
-    GlobalExecutor::executeMemberOnDereferenced(
-        &LocalAssemblerInterface::preTimestep, _local_assemblers,
-        *_local_to_global_index_map, x, t, dt);
+    // If monolithic scheme is used or the equation of deformation is solved in
+    // the staggered scheme.
+    if (_use_monolithic_scheme || process_id == 1)
+        GlobalExecutor::executeMemberOnDereferenced(
+            &LocalAssemblerInterface::preTimestep, _local_assemblers,
+            *_local_to_global_index_map, x, t, dt);
 }
 
 template <int DisplacementDim>
 void HydroMechanicsProcess<DisplacementDim>::postTimestepConcreteProcess(
-    GlobalVector const& x)
+    GlobalVector const& x, const int process_id)
 {
     DBUG("PostTimestep HydroMechanicsProcess.");
-
     GlobalExecutor::executeMemberOnDereferenced(
         &LocalAssemblerInterface::postTimestep, _local_assemblers,
-        *_local_to_global_index_map, x);
+        getDOFTable(process_id), x);
+}
+
+template <int DisplacementDim>
+void HydroMechanicsProcess<DisplacementDim>::postNonLinearSolverProcess(
+    GlobalVector const& x, const double t, const int process_id)
+{
+    if (!_use_monolithic_scheme && process_id == 0)
```

Comment:
Maybe there should be a method `! isLastProcess()` for this condition.

Reply:
Introduced a new member:  
`bool HydroMechanicsProcess::hasMechanicalProcess(int const process_id) `.

Reply:
OK, for some occasions the usage of this method is more mechanics related (e.g. in getDOFtable), and sometimes it is more is-last-process related (e.g. in post timestep).